Add button to view @ mentions.

sumanthvrao opened the pull request:
This PR adds support to view `@` mentioned messages. Any feedback regarding this is welcomed.

Solves #425

amanagr commented:
One thing I noticed is that if there are two message of the same narrow (say #zt>some topic), and if I narrow using  one of them via `S/z` from the `@mention` narrow to `some topic` and come back to `@mention` narrow and select the other msg of the same narrow (`some topic`) and press `z`, it's focus is still set on the previous message. This might come as a separate PR btw.

sumanthvrao commented:
@amanagr Not sure why, But I'm not able to reproduce this. How are you coming back to the `@mention` narrow? (By selecting it from the left sidebar?). 

My steps:
* Went to `@mention` narrow and clicked on `#checkin>sumanth` which has multiple mentioned messages. I clicked on the top and used `z`, which zoomed into Checkin stream
* I again went to `@mention` narrow choosing from side bar and did the same on the other message. Which took me to the correct focus.

Am I doing anything wrong?

amanagr commented:
I am also unable to reproduce it now :|

sumanthvrao commented:
@neiljp FYI. Updated this with tests. Its now ready for a review.

sumanthvrao commented:
@neiljp Resolved conflicts and rebased this. Do you have time to give it a review?

sumanthvrao commented:
This PR has be split up as :

* The first 3 commits adds the necessary support functions involving the new 'mentioned messages' narrow; These involve - create narrow, get id's in narrow and show all messages in the narrow. Tests have been added in the respective commits as well.
* The next 3 commits are UI related. They help register the new hotkey (`#`) , connect it via keypress and change the search bar text as well.
* The 7th and 8th commit - create the UI button for the 'mentioned messages' and connect it via the UI.
* The last commit classifies the unread messages so that the UI button can be marked with them.

I have separated the commits which involve the scenario when a mentioned message is received after ZT has started. I have kept these in a separate branch and will work on these shortly.

@neiljp @amanagr

neiljp commented on zulipterminal/config/keys.py:
```diff
@@ -140,6 +140,11 @@
     }),
+    ('ALL_MENTIONS', {
+        'keys': {'#'},
+        'help_text': 'Narrow to all mentioned messages',
```
Sync this with the README, of course :)

neiljp commented on zulipterminal/ui.py:
```diff
@@ -164,6 +164,9 @@ def keypress(self, size: Tuple[int, int], key: str) -> Optional[str]:
             self.body.focus_col = 1
+        elif is_command_key('ALL_MENTIONS', key):
+            self.model.controller.show_all_mentions(self)
+            self.body.focus_col = 1
```
Perhaps a little minor, but perhaps squash this and the README/keys commit, to avoid a potential state where the docs/help don't match the UI (even if the visual button doesn't exist, the key does work!)

sumanthvrao replied:
Good catch!

neiljp commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -118,6 +118,18 @@ def __init__(self, controller: Any, width: int) -> None:
+                         controller.show_all_mentions,
+                         width=width,
+                         count=count,
+                         prefix_character='')
```
There's no test for this, but this seems reasonable for now given the simplicity :)

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -665,10 +665,16 @@ def menu_view(self) -> Any:
             self.view.home_button,
             self.view.pm_button,
             self.view.starred_button,
+            self.view.mentioned_button,
```
Perhaps we should switch the order here, to match the webapp, at least for now?
(the order of the `MentionedButton` code relative to the `StarredButton` too, if so)

sumanthvrao replied:
Sure. I changed the order of the buttons and code in respective places.

sumanthvrao commented:
@neiljp Thanks for the review. I have updated the PR with the changes requested.

neiljp commented on README.md:
```diff
@@ -116,6 +116,7 @@ Install-Module -Name BurntToast
 | Go to the last message                                | <kbd>G</kbd> / <kbd>end</kbd>                 |
 | Narrow to all private messages                        | <kbd>P</kbd>                                  |
 | Narrow to all starred messages                        | <kbd>f</kbd>                                  |
+| Narrow to all mentioned mentions                      | <kbd>#</kbd>                                  |
```
I think we lost the `Narrow to` prefix in the latest version; did you mean to?

neiljp commented on zulipterminal/core.py:
```diff
@@ -233,6 +233,12 @@ def show_all_starred(self, button: Any) -> None:
                         starred=True)
 
+    def show_all_mentions(self, button: Any) -> None:
+        # NOTE: Should we ensure we maintain anchor focus here?
```
It *seems* to maintain focus so far without? However as with 'starred', perhaps add a note to indicate that, as we may benefit from it later?

sumanthvrao replied:
Added 👍

neiljp commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -106,6 +106,18 @@ def __init__(self, controller: Any, width: int, count: int=0) -> None:
 
+class MentionedButton(TopButton):
+    def __init__(self, controller: Any, width: int, count: int=0) -> None:
+        button_text = ("Mentions [" +
```
Maybe we should adjust the spacing to look a little better, hard-coded for now?

sumanthvrao replied:
Anything particular in mind?

sumanthvrao commented:
@neiljp What do you want the README/help text to say. I think `Messages in which you're mentioned` seems good.

The new commit message was because I squashed 2 commits from the previous version. I have updated the commit heading to be the first one of the squashed  commits.

I didn't exactly get what changes you wanted wrt button spacing. 

Besides this, if there are no more changes, let's get it merged ASAP :)

neiljp commented:
@sumanthvrao The first 5 are merged; now onto more UI :) :tada: 

I think the other work is updating while ZT is running, and possibly the all-mentions aspect?

sumanthvrao commented:
Thanks @neiljp !

sumanthvrao commented:
Thanks for the review @kaustubh-nair! Glad to see there are no errors. :)

neiljp commented:
@sumanthvrao I just merged the rest of these commits manually, after slightly adjusting the last one to fit with the recent changes to `set_count` - working well so far, thanks! :tada:

Some commits are quite small, but we put in a lot of related effort before this - though some extra tests may be warranted?

preetmishra commented:
@sumanthvrao Awesome work! This looks great. :+1:

sumanthvrao commented:
@neiljp Thanks for taking time to review this and I'm glag to finally get it merged. 

Thanks @preetmishra :)
